### PR TITLE
Refactor writing Win/Unix line endings

### LINKED ISSUE #989: .Rd files are forcibly rewritten with LF ending on Windows
despite of what's announced in the NEWS and in #841.

To reproduce:

- Clone any package on Windows
- Run `unix2dos man/*.Rd`
- Run `devtools::document()`

I believe this is because `write_if_different()` already receives character strings with embedded newlines, and `write_lines()` doesn't convert them to the detected newlines.

### PR DESCRIPTION
**Pull request by gaborcsardi**
We need to move the detection of line endings
upstream a bit, to be able to tell if the
newly generated contents is different from the
current one, in `write_if_different()`.

This means that we need to `paste()` the contents
together in `write_if_different()`. Consequently,
we might as well remove the `paste()`-ing from
`same_contents()`, as `write_if_different()` always
calls it with the `paste()`-d contents.

Also, if we `paste()` already upstream, then
we might as well use `writeBin()` to write out
the contents, instead of `write_lines()`, especially
because `write_lines()` would add an extra empty line.
`write_lines()` is still kept, and used in some
code paths, and it can also do EOL-detection,
although this feature is not used currently.

Another benefit of this refactoring is that we
also move `enc2utf8()` upstream, this should clearly
happen before we check if the new contents is
different.

Added test cases as well.

Fixes #989, supersedes #1003.

### CODE REVIEW

File: R/utils-io.R [hunk 1]
```diff
@@ -5,13 +5,18 @@ read_lines <- function(path, n = -1L) {
...
 }
 
-write_lines <- function(text, path) {
-  if (has_windows_le(path)) {
-    line_ending <- "\r\n"
-  } else {
-    line_ending <- "\n"
+write_lines <- function(text, path, line_ending = NULL) {
+  if (is.null(line_ending)) {
+    if (has_windows_le(path)) {
```
Comment (krlmlr) [inline R/utils-io.R]: Extract `get_le_from_path()` ?

File: R/utils.R [hunk 2]
```diff
@@ -73,6 +73,14 @@ write_if_different <- function(path, contents, check = TRUE) {
...
   }
 
+  if (has_windows_le(path)) {
+    line_ending <- "\r\n"
+  } else {
+    line_ending <- "\n"
+  }
+
+  contents <- paste0(paste0(contents, collapse = line_ending), line_ending)
+  contents <- gsub("\r?\n", line_ending, contents)
```
Comment (krlmlr) [inline R/utils.R]:
```suggestion
  contents <- enc2utf8(gsub("\r?\n", line_ending, contents))
```

and remove the conversion below?

Comment (gaborcsardi) [reply R/utils.R]: Right, good catch, fixed now.

File: R/utils-io.R (hunk 1 above)
Comment (gaborcsardi) [reply R/utils-io.R]: That's an existing function, so I don't want to rename it.

**Comment by gaborcsardi** [comment]
@krlmlr Thanks!

**Comment by gaborcsardi** [comment]
> Maybe we should have a function that returns the line ending instead of a function that returns a logical if it’s a windows line ending? I think that would reduce a little duplication. And maybe make same_contents error of contents isn’t length 1?

I'll do these, I was considering them as well.

**Comment by gaborcsardi** [comment]
@hadley OK, I did both, and added NEWS. I'll wait for the CI checks, and then merge.

**Pull request merged**